[CC-3025] sqlproxy: rate limit proxy connections by IP

spaskob opened the pull request:
Add a new admitter package which performs basic connection admission
control to rate limit connection attempts. Connection attempts are
rate limited based on source IP. Admission control is currently
purely a local decision, but the package this interface defines could
easily become the client-side for connection to a centralized service.

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56192)
<!-- Reviewable:end -->

tbg commented:
@asubiotto could you or someone else on the SQL end review this?

petermattis reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> for the `admitter` code which was forked from code in CC land. The `sqlproxyccl` changes look reasonable enough, though I've never looked at that code before. I think there should be a test that this rate limiting is working as expected. In CC code, we utilize the the `admitter.Service` mock to force various code paths during login. I can see a similar test being implemented for `sqlproxyccl`

Looks like `c-deps/{cryptopp,googletest}` were re-added accidentally. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-:-MLDFfwlBTMFDkwZSMYU:b-yzf7o9)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @spaskob and @tbg)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLDFJWH-8cFKYC8H56L:-MLDFJWH-8cFKYC8H56M:b-s1f58x) ([raw file](https://github.com/cockroachdb/cockroach/blob/f373eb953bd93b22c0f0fdf8390d228e9b7f524b/pkg/ccl/sqlproxyccl/proxy.go#L59)):*
> ```Go
> 
> 	if s.admitter != nil {
> 		// TODO(spaskob): check for previous successful connection from the same IP
> ```

I wonder if we'll have to do this soon. Connections to the SQL proxy are likely much more frequent that console logins. 


<!-- Sent from Reviewable.io -->

spaskob reviewed (commented):
FYI I will be adding the recommended test soon.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-:-MLDXujz3IwVmpxcuRyR:b-7pkgot)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @petermattis and @tbg)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLDFJWH-8cFKYC8H56L:-MLDWQZc2_AEO8m1Z4as:b-ea4y2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/f373eb953bd93b22c0f0fdf8390d228e9b7f524b/pkg/ccl/sqlproxyccl/proxy.go#L59)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I wonder if we'll have to do this soon. Connections to the SQL proxy are likely much more frequent that console logins. 
</blockquote></details>

Let me try to add something even in this PR.
Some clarifying questions:

1. Is metadata about successful connections persisted somewhere?
2. Do we need to take into an account the tenant id as well? A hacker could connect to their own tenant and then use the free-pass to attack others.
3. The above may already be a vulnerability of the current rate limiting approach. Perhaps we should track simultaneous connections from the same IP?


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @spaskob and @tbg)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLDFJWH-8cFKYC8H56L:-MLDdoZN8mw9LhrvUKoM:bg4n7hl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f373eb953bd93b22c0f0fdf8390d228e9b7f524b/pkg/ccl/sqlproxyccl/proxy.go#L59)):*
> 1. Is metadata about successful connections persisted somewhere?

Not that I know of. I suspect we don't want to persist it as it is nice for `sqlproxy` to be stateless. Perhaps caching it in memory is sufficient. I'm imagining a cache keyed by tenant ID and source IP address. 

> 2. Do we need to take into an account the tenant id as well? A hacker could connect to their own tenant and then use the free-pass to attack others.

Yes. We definitely need to account for the tenant ID.

> 3. The above may already be a vulnerability of the current rate limiting approach. Perhaps we should track simultaneous connections from the same IP?

Can you elaborate on this?


<!-- Sent from Reviewable.io -->

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @petermattis and @tbg)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLDFJWH-8cFKYC8H56L:-MLJkojvD0-Y2iMciThS:bg9en82) ([raw file](https://github.com/cockroachdb/cockroach/blob/f373eb953bd93b22c0f0fdf8390d228e9b7f524b/pkg/ccl/sqlproxyccl/proxy.go#L59)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

> 1. Is metadata about successful connections persisted somewhere?

Not that I know of. I suspect we don't want to persist it as it is nice for `sqlproxy` to be stateless. Perhaps caching it in memory is sufficient. I'm imagining a cache keyed by tenant ID and source IP address. 

> 2. Do we need to take into an account the tenant id as well? A hacker could connect to their own tenant and then use the free-pass to attack others.

Yes. We definitely need to account for the tenant ID.

> 3. The above may already be a vulnerability of the current rate limiting approach. Perhaps we should track simultaneous connections from the same IP?

Can you elaborate on this?
</blockquote></details>

1. Caching is being added (still WIP).
2. Talking to Jay on how to  actually parse the tenant id from the connection params. It is doable - we are figuring out the details.

As for 3:

Is this scenario possible:

- connection attempts to tenant X from a given IP get throttled
- so attacker successfully connects to their own tenant Y through the same IP instead
- this clears the IP from the admitter
- the attacker resumes the attack on tenant X


<!-- Sent from Reviewable.io -->

asubiotto commented:
@tbg I see you removed a request for review, do you still want me to take a look at this?

tbg commented:
Yes, sorry! I must have fat fingered it.

asubiotto reviewed (dismissed):
Would be good to simulate the single IP multiple tenant ID scenario you point out in the test you're adding.

> Reviewed 2 of 8 files at r1, 7 of 7 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-:-MLNHuK6CnnIA3D3B-81:b-jg078j)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @petermattis, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 42 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNHLOr9kPyZEix06vY:-MLNHLOsEsjmIJj_wubE:b-6jcb72) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/proxy.go#L42)):*
> ```Go
> 	OnSendErrToClient func(code ErrorCode, msg string) string
> 
> 	admitter admitter.Service
> ```

nit: as far as I can see, it doesn't seem like this field is used anywhere (the one on the `Server` is)
___
*[pkg/ccl/sqlproxyccl/admitter/local.go, line 114 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNJ_cQ3Irsg0ym__0r:-MLNJ_cQ3Irsg0ym__0s:bx7tc6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f1411807a0a1827e912b1c392ea0ff70713a866/pkg/ccl/sqlproxyccl/admitter/local.go#L114)):*
> ```Go
> 		index: len(s.mu.limiters),
> 	}
> 	s.mu.limiters[addr] = l
> ```

can you reset and reuse the limiter rather than allocating a new one? does it make sense to do so? Just realizing now this is a fork of existing code, feel free to ignore this comment.
___
*[pkg/ccl/sqlproxyccl/admitter/local.go, line 79 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNISrj2wNg98xddtDq:-MLNISrj2wNg98xddtDr:bhzdw2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/admitter/local.go#L79)):*
> ```Go
> 	}
> 	s.mu.limiters = make(map[string]*limiter)
> 	s.mu.knownTenants = make(map[string]map[uint64]bool)
> ```

If existence is enough, consider making this a `map[string]map[uint64]struct{}` to use less memory?
___
*[pkg/ccl/sqlproxyccl/admitter/local.go, line 109 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNKwPWAHFXkihfzB3c:-MLNKwPX-WyfSCE_4JnO:brhumzt) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/admitter/local.go#L109)):*
> ```Go
> 		s.mu.knownTenants[ipAddress] = map[uint64]bool{tenID: true}
> 	} else {
> 		s.mu.knownTenants[ipAddress][tenID] = true
> ```

Same here.
___
*[pkg/ccl/sqlproxyccl/admitter/local.go, line 116 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNKfkYALeEUcK-Bq6e:-MLNKfkYALeEUcK-Bq6f:b-ks5223) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/admitter/local.go#L116)):*
> ```Go
> 	s.mu.Lock()
> 	defer s.mu.Unlock()
> 	_, ok := s.mu.knownTenants[ipAddress]
> ```

You can avoid a second lookup by storing the result of this lookup and reusing the map in the `if ok` block.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @petermattis, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 123 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNTuKx-RTJo4B97wTu:-MLNTuKx-RTJo4B97wTv:benfo4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/proxy.go#L123)):*
> ```Go
> 	if s.admitter != nil {
> 		ip := conn.RemoteAddr().String()
> 		tenID, clientErr = s.opts.TenantFromParams(msg.Parameters)
> ```

I believe this is the first introduction of the concept of tenants to the proxy code. I think it would be better to keep tenant logic separated. One way to do this is to define a local `Admitter` interface that takes the IP address and `msg.Parameters`:

```
type Admitter interface {
  AllowRequest(ip string, params map[string]string], ...) error
}
```
___
*[pkg/ccl/sqlproxyccl/admitter/local.go, line 52 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNV04eDg22j9i3MAfQ:-MLNV04f6QIFAN8ZRUBR:b-yz1yit) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/admitter/local.go#L52)):*
> ```Go
> 		limiters map[string]*limiter
> 		// Map from IP address to known tenant ids for this IP.
> 		knownTenants map[string]map[uint64]bool
> ```

I think you're always looking up the tuple `(ip-address, tenant-id)`. If that is the case, it would be faster and use less memory to have a single level map:

```
type tenantKey struct {
  ipAddress string
  tenantID uint64
}
...
knownTenants map[tenantKey]struct{}
```

Every map has overhead. This approach switches from a map per ip address to a single map which should be a significant reduction in memory.

Related to my other comment: this is the first introduction of the concept of tenants into the SQL proxy code. I think we should avoid it. The `knownTenants` stuff can be pulled out by adjusting the admitter interface used by SQL proxy and having the caller implement the known tenant logic. Specifically, my suggestion is that the known tenant handling get pulled into a separate struct and implemented in the managed-service repo.

Do we need a max size on this map? I think we can get away with not having one because the `knownTenants` map is limited by the number of tenants and how many IPs they are successfully be connected from. If that sounds right to you, it is worthwhile putting that explanation in a comment.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aaron-crl, @asubiotto, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLDFJWH-8cFKYC8H56L:-MLNWoWy12VYTcX6QX0C:b-841ykd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f373eb953bd93b22c0f0fdf8390d228e9b7f524b/pkg/ccl/sqlproxyccl/proxy.go#L59)):*
> Is this scenario possible:

This sounds possible, but unlikely, and not too harmful. The attacker needs to be able to spoof an IP address that someone is using log in from. How do they know that IP? Assuming they somehow do know such an IP, all they are getting is periodic resets of the rate limiter. Most of their attempts would still be throttled unless the successful connections from that IP address are happening very frequently.

Cc @aaron-crl for his thoughts.


<!-- Sent from Reviewable.io -->

aaron-crl reviewed (commented):
This probably ought to have a release note as enabling this can result in users locking themselves out of a cluster in a way previously not possible.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-:-MLNpVAzAEOefE54m3D-:bdi6pv8)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @petermattis, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLDFJWH-8cFKYC8H56L:-MLNqSwl4DWkMjl3jGDd:b-g5cgh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f373eb953bd93b22c0f0fdf8390d228e9b7f524b/pkg/ccl/sqlproxyccl/proxy.go#L59)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

> Is this scenario possible:

This sounds possible, but unlikely, and not too harmful. The attacker needs to be able to spoof an IP address that someone is using log in from. How do they know that IP? Assuming they somehow do know such an IP, all they are getting is periodic resets of the rate limiter. Most of their attempts would still be throttled unless the successful connections from that IP address are happening very frequently.

Cc @aaron-crl for his thoughts.
</blockquote></details>

I looked at the local service code in this PR and I'm not sure I understand the proposed attack.

Would this be the attacker loop?
* Attacker makes _n_ guesses against `target_tenant` from `attacker_IP`
* Attacker signs into `attacker_tenant` reseting the counter from `attacker_IP`

It looks like we track both the tenant and the IP in the knownTenant map so this wouldn't unblock attempts from the attacker's IP for `target_tenant` .

Is the concern that the attacker could spoof a valid connection from the legitimate user's IP? That should be virtually impossible if authentication takes place under TLS.


<!-- Sent from Reviewable.io -->

spaskob commented:
> This probably ought to have a release note as enabling this can result in users locking themselves out of a cluster in a way previously not possible.
> 
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-:-MLNpVAzAEOefE54m3D-:bdi6pv8)_ status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @petermattis, and @spaskob)
> 
> _[pkg/ccl/sqlproxyccl/proxy.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLDFJWH-8cFKYC8H56L:-MLNqSwl4DWkMjl3jGDd:b-g5cgh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f373eb953bd93b22c0f0fdf8390d228e9b7f524b/pkg/ccl/sqlproxyccl/proxy.go#L59)):_
> 
> Previously, petermattis (Peter Mattis) wrote…
> I looked at the local service code in this PR and I'm not sure I understand the proposed attack.
> 
> Would this be the attacker loop?
> 
> * Attacker makes _n_ guesses against `target_tenant` from `attacker_IP`
> * Attacker signs into `attacker_tenant` reseting the counter from `attacker_IP`
> 
> It looks like we track both the tenant and the IP in the knownTenant map so this wouldn't unblock attempts from the attacker's IP for `target_tenant` .
> 
> Is the concern that the attacker could spoof a valid connection from the legitimate user's IP? That should be virtually impossible if authentication takes place under TLS.

Yes exactly this is what I am worried about.
`knownTenant` map is used to bypass the admitter entirely - if `(IP,tenID)` has had a known successful login, we don't try to rate limit connection request.
The `admitter` however only tracks requests by IP, hence it is possible for the attacker to reset their exp backoff to 0 on `target_tenant` by successfully connecting from their same IP to their own `attacker_tenant`.

aaron-crl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @petermattis, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLDFJWH-8cFKYC8H56L:-MLOMgvy8fOwWxwpYL64:b-vrmlw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f373eb953bd93b22c0f0fdf8390d228e9b7f524b/pkg/ccl/sqlproxyccl/proxy.go#L59)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

Yes exactly this is what I am worried about. `knownTenant` map is used to bypass the admitter entirely - if `(IP,tenID)` has had a known successful login, we don't try to rate limit connection request. The `admitter` however only tracks requests by IP, hence it is possible for the attacker to reset their exp backoff to 0 on `target_tenant` by successfully connecting from their same IP to their own `attacker_tenant`. 
</blockquote></details>

Ah! Good find! Yes, we should protect against that case. I missed that on the original review.

Impact is that any user with valid credentials can circumvent rate-limiting protection for another user account (which is mercifully limited).

We still want to protect against the base case of an an attacker attempting to check for common passwords against multiple accounts, but we should _also_ limit at the user/tenant, IP level (but not at the user/tenant level to limit DDoS channels).


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aaron-crl, @asubiotto, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLDFJWH-8cFKYC8H56L:-MLP9CCeAS0ZkW7eU85T:bpb1q5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/f373eb953bd93b22c0f0fdf8390d228e9b7f524b/pkg/ccl/sqlproxyccl/proxy.go#L59)):*
<details><summary><i>Previously, aaron-crl wrote…</i></summary><blockquote>

Ah! Good find! Yes, we should protect against that case. I missed that on the original review.

Impact is that any user with valid credentials can circumvent rate-limiting protection for another user account (which is mercifully limited).

We still want to protect against the base case of an an attacker attempting to check for common passwords against multiple accounts, but we should _also_ limit at the user/tenant, IP level (but not at the user/tenant level to limit DDoS channels).
</blockquote></details>

Interesting. Is it sufficient to get rid of the `RequestSuccess` path? The `knownTenant` check will disable rate limiting for a particular tenant once they successfully auth, but we'd never reset the rate limit for a particular IP address. 


<!-- Sent from Reviewable.io -->

spaskob commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192)_ status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aaron-crl, @asubiotto, and @spaskob)
> 
> _[pkg/ccl/sqlproxyccl/proxy.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLDFJWH-8cFKYC8H56L:-MLP9CCeAS0ZkW7eU85T:bpb1q5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/f373eb953bd93b22c0f0fdf8390d228e9b7f524b/pkg/ccl/sqlproxyccl/proxy.go#L59)):_
> 
> Previously, aaron-crl wrote…
> Interesting. Is it sufficient to get rid of the `RequestSuccess` path? The `knownTenant` check will disable rate limiting for a particular tenant once they successfully auth, but we'd never reset the rate limit for a particular IP address.

Yes - that should work. I checked the existing throttling code in CC and it seems that we should do the same there. WDYT?

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @petermattis, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 123 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNTuKx-RTJo4B97wTu:-MLSSmd95QBG3aaEzU7v:bhs3a2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/proxy.go#L123)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I believe this is the first introduction of the concept of tenants to the proxy code. I think it would be better to keep tenant logic separated. One way to do this is to define a local `Admitter` interface that takes the IP address and `msg.Parameters`:

```
type Admitter interface {
  AllowRequest(ip string, params map[string]string], ...) error
}
```
</blockquote></details>

Now that I understand better the
___
*[pkg/ccl/sqlproxyccl/admitter/local.go, line 52 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNV04eDg22j9i3MAfQ:-MLSSt7E3DiJ1HsI-ozK:b-ta2g5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/admitter/local.go#L52)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think you're always looking up the tuple `(ip-address, tenant-id)`. If that is the case, it would be faster and use less memory to have a single level map:

```
type tenantKey struct {
  ipAddress string
  tenantID uint64
}
...
knownTenants map[tenantKey]struct{}
```

Every map has overhead. This approach switches from a map per ip address to a single map which should be a significant reduction in memory.

Related to my other comment: this is the first introduction of the concept of tenants into the SQL proxy code. I think we should avoid it. The `knownTenants` stuff can be pulled out by adjusting the admitter interface used by SQL proxy and having the caller implement the known tenant logic. Specifically, my suggestion is that the known tenant handling get pulled into a separate struct and implemented in the managed-service repo.

Do we need a max size on this map? I think we can get away with not having one because the `knownTenants` map is limited by the number of tenants and how many IPs they are successfully be connected from. If that sounds right to you, it is worthwhile putting that explanation in a comment.
</blockquote></details>

Thanks for the detailed explanation, the purpose of the proxy makes much more sense to me now. Should we take your argument even further and pass the admitter and `knownClient` cache as options in `sqlproxyccl.Options` https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/sqlproxyccl/proxy.go#L26 ? Thus we won't have to fork the admitter library at all.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aaron-crl, @asubiotto, @petermattis, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLDFJWH-8cFKYC8H56L:-MLS_8PHCzZNt7Vg1tor:b-x6cu2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/f373eb953bd93b22c0f0fdf8390d228e9b7f524b/pkg/ccl/sqlproxyccl/proxy.go#L59)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

Yes - that should work. I checked the existing throttling code in CC and it seems that we should do the same there. WDYT? 
</blockquote></details>

Agreed, we should fix the CC code as well. Can you take care of that?
___
*[pkg/ccl/sqlproxyccl/admitter/local.go, line 52 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNV04eDg22j9i3MAfQ:-MLS_JyP1OpxfbRfYAta:bjghzrs) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/admitter/local.go#L52)):*
> Should we take your argument even further and pass the admitter and knownClient cache as options in sqlproxyccl.Options https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/sqlproxyccl/proxy.go#L26 ? Thus we won't have to fork the admitter library at all.

This is a good idea. Are you imagining defining an `Admitter` interface as I mentioned in another message? I think that would work. Another possibility is that we extend `OutgoingAddrFromParams` (which is now called `BackendFromParams` after Jackson's recent PR). Seems like any throttling could be hidden in that callback. We'd just have to also supply the remote IP address.


<!-- Sent from Reviewable.io -->

aaron-crl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @petermattis, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLDFJWH-8cFKYC8H56L:-MLSk2wl-CCmITX3_M_r:b-d2z8cq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f373eb953bd93b22c0f0fdf8390d228e9b7f524b/pkg/ccl/sqlproxyccl/proxy.go#L59)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Agreed, we should fix the CC code as well. Can you take care of that?
</blockquote></details>

Sounds good on all counts.


<!-- Sent from Reviewable.io -->

spaskob reviewed (commented):
Thanks for the comments - based on them I am going to change this PR significantly and ping you for further reviews. @asubiotto I need guidance on how to unskip this test https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/sqlproxyccl/proxy_test.go#L205 so that I can use it as template for the test testing the throttling logic.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-:-MLSvzFE3JzVK_Y4pR5E:bbbicm5)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @petermattis, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLDFJWH-8cFKYC8H56L:-MLSv7lM5jO84HwO92Ef:bhy9oss) ([raw file](https://github.com/cockroachdb/cockroach/blob/f373eb953bd93b22c0f0fdf8390d228e9b7f524b/pkg/ccl/sqlproxyccl/proxy.go#L59)):*
<details><summary><i>Previously, aaron-crl wrote…</i></summary><blockquote>

Sounds good on all counts.
</blockquote></details>

Will do 
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 42 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNHLOr9kPyZEix06vY:-MLSuthU5HevYpM5TT0_:b-1oas8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/proxy.go#L42)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: as far as I can see, it doesn't seem like this field is used anywhere (the one on the `Server` is)
</blockquote></details>

ack - we are taking an approach not using a explicit admitter instance - this will go away.
___
*[pkg/ccl/sqlproxyccl/admitter/local.go, line 114 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNJ_cQ3Irsg0ym__0r:-MLSvued6jXNeyl2sofo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f1411807a0a1827e912b1c392ea0ff70713a866/pkg/ccl/sqlproxyccl/admitter/local.go#L114)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

can you reset and reuse the limiter rather than allocating a new one? does it make sense to do so? Just realizing now this is a fork of existing code, feel free to ignore this comment.
</blockquote></details>

Done.
___
*[pkg/ccl/sqlproxyccl/admitter/local.go, line 52 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNV04eDg22j9i3MAfQ:-MLSvQ6u0QjUPOfQ7KE2:b-7cumy2) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/admitter/local.go#L52)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

> Should we take your argument even further and pass the admitter and knownClient cache as options in sqlproxyccl.Options https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/sqlproxyccl/proxy.go#L26 ? Thus we won't have to fork the admitter library at all.

This is a good idea. Are you imagining defining an `Admitter` interface as I mentioned in another message? I think that would work. Another possibility is that we extend `OutgoingAddrFromParams` (which is now called `BackendFromParams` after Jackson's recent PR). Seems like any throttling could be hidden in that callback. We'd just have to also supply the remote IP address.
</blockquote></details>

I like the latter better now that you mention it, so that we don't have to expose the existence of an admitter at all.
___
*[pkg/ccl/sqlproxyccl/admitter/local.go, line 79 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNISrj2wNg98xddtDq:-MLSvjkpCd_Lmlj01urP:b-1oas8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/admitter/local.go#L79)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

If existence is enough, consider making this a `map[string]map[uint64]struct{}` to use less memory?
</blockquote></details>

ack - we are taking an approach not using a explicit admitter instance - this will go away.
___
*[pkg/ccl/sqlproxyccl/admitter/local.go, line 109 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNKwPWAHFXkihfzB3c:-MLSvnYw7E0p7J4J0q1E:b-1oas8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/admitter/local.go#L109)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Same here.
</blockquote></details>

ack - we are taking an approach not using a explicit admitter instance - this will go away.
___
*[pkg/ccl/sqlproxyccl/admitter/local.go, line 116 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNKfkYALeEUcK-Bq6e:-MLSvrmz9t-Y6qmpTRyu:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/admitter/local.go#L116)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

You can avoid a second lookup by storing the result of this lookup and reusing the map in the `if ok` block.
</blockquote></details>

ditto


<!-- Sent from Reviewable.io -->

ajwerner reviewed (dismissed):
Seems like you need a bazel regen. The more we find ourselves copying code rather than sharing and improving, the sadder I'll be. That doesn't need to be solved for this but it's something we should be mindful of. 


I'd love to see some testing of

* The KnownClient logic 
* The plumbing in the last commit, can you write something of an integration test that ensures clients get the appropriate error in the appropriate cases? 

> Reviewed 1 of 8 files at r1, 1 of 7 files at r3, 3 of 8 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-:-MLiD4bp535e9o6eIaAh:b8c7aqp)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @petermattis, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/admitter/service.go, line 18 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLiEW-b8UypbOZF1bV1:-MLiEW-b8UypbOZF1bV2:bfie2rg) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/admitter/service.go#L18)):*
> ```Go
> 	// AllowRequest determines whether a request should be allowed to proceed. It
> 	// rate limits requests from IP addresses regardless of tenant id.
> 	AllowRequest(ipAddress string, now time.Time) error
> ```

Returning an error here seems like too big of a hammer. Given you aren't providing any guidance on the set of errors being returned even though it's just one, I think returning  a bool would make for a better API. 
___
*[pkg/ccl/sqlproxyccl/admitter/service.go, line 23 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLiEpFp9KFgawt6vDfg:-MLiEpFp9KFgawt6vDfh:b-leyxjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/admitter/service.go#L23)):*
> ```
> 	// KnownClient checks if this client has connected successfully before.
> 	KnownClient(ipAddress string, tenID uint64) bool
> ```
Is it definitely better to have this as a separate method and force the caller to check this before calling `AllowRequest`? Should we just augment the `AllowRequest` interface to also take the `tenID` and short-circuit?


<!-- Sent from Reviewable.io -->

petermattis commented on pkg/ccl/sqlproxyccl/proxy.go:
```diff
@@ -108,15 +109,32 @@ func (s *Server) Proxy(conn net.Conn) error {
 	}
 
-	outgoingAddr, outgoingTLS, clientErr := s.opts.BackendFromParams(msg.Parameters)
+	backendConfig, clientErr := s.opts.BackendConfigFromParams(msg.Parameters)
```
Rather than returning an admitter to check in `backendConfig.Admitter`, could you instead pass `ipAddress` as an additional argument to `BackendConfigFromParams` and do the IP-based rate limiting within that function itself?

spaskob replied:
Sure, should we also hide the caching of known connections in this function as well then? Perhaps `BackendConfigFromParams` can additionally return a callback stored in `BackendConfig`, say `func CacheConn() error`, to be called after successful login? This will replace the need to provide a `ClientID` that is cached by the proxy.

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @petermattis, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 112 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-ML7Wk4mqJ6wzlfGHwA4-r5-112:-MLmGXdXBBo1X8_uHoHb:b-si0vxh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f079b2cb76155bcea7427fd5882afe6b5ba4d53/pkg/ccl/sqlproxyccl/proxy.go#L112)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

Sure, should we also hide the caching of known connections in this function as well then? Perhaps `BackendConfigFromParams` can additionally return a callback stored in `BackendConfig`, say `func CacheConn() error`, to be called after successful login? This will replace the need to provide a `ClientID` that is cached by the proxy.
</blockquote></details>

Yes, that approach sounds good, though I have a nit about the naming. The "known tenants" cache is caching whether a tenant connection attempt was successful. Naming it `CacheConn` makes me think that it is caching the connection itself. I'd name this callback something like `ConnectionSuccess`.


<!-- Sent from Reviewable.io -->

Commit:
[CC-3025] sqlproxy: add hook to rate limit connections

We need a mechanism for sqlproxy users to provide admission control to
clients that is enforced in the proxy. To that end we expand
`BackendConfigFromParams` to provide a signal whether a connection
should be refused and a callback to record the success of a connection.
The callback can implement caching of known connections for example so
that they are not rate limited on further connection requests.

Release note: none.

spaskob commented:
PTAL

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @petermattis, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/metrics.go, line 76 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLrHfvbB24GMKfIK8VA:-MLrHfvbB24GMKfIK8VB:bj1m0kc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5055720eab4f00ace41f8879997fea8c6142788/pkg/ccl/sqlproxyccl/metrics.go#L76)):*
> ```Go
> 		CurConnCount:           metric.NewGauge(metaCurConnCount),
> 		RoutingErrCount:        metric.NewCounter(metaRoutingErrCount),
> 		RefusedConnCount:       metric.NewCounter(metaBackendDisconnectCount),
> ```

`s/metaBackendDisconnectCount/metaRefusedConnCount/g`
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 28 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLrHE4U2RghXOr3aE_-:-MLrHE4U2RghXOr3aE_0:b-9tf2en) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5055720eab4f00ace41f8879997fea8c6142788/pkg/ccl/sqlproxyccl/proxy.go#L28)):*
> ```Go
> // being proxied.
> type BackendConfig struct {
> 	OutgoingAddress     string
> ```

Let's document these fields.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 31 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLrJp0uDo28X4Xj_dbd:-MLrJp0uDo28X4Xj_dbe:bhzsyge) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5055720eab4f00ace41f8879997fea8c6142788/pkg/ccl/sqlproxyccl/proxy.go#L31)):*
> ```Go
> 	TLSConf             *tls.Config
> 	RefuseConn          bool
> 	OnConnectionSuccess func() error
> ```

Under what circumstances would this callback want to return an error?
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 133 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLrI4WCCUM70DL6Vo6z:-MLrI4WCCUM70DL6Vo7-:b-h2revu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5055720eab4f00ace41f8879997fea8c6142788/pkg/ccl/sqlproxyccl/proxy.go#L133)):*
> ```Go
> 		var clientErr error
> 		backendConfig, clientErr = s.opts.BackendConfigFromParams(msg.Parameters, ip)
> 		if clientErr != nil {
> ```

Rather than the special `backendConfig.RefuseConn` handling, perhaps we should export `newErrorf` and then allow inspecting it here. That is, we'd do something like:

```
if clientErr != nil {
  s.metrics.RoutingErrCount.Inc(1)

  var codeErr *codeError
  if !errors.As(clientErr, &codeErr) {
    codeErr = newErrorf(CodeParamsRoutingFailed, "rejected by BackendConfigFromParams: %v", clientErr)
  }
  sendErrToClient(conn, codeErr.code, clientErr.Error())
  return codeErr
}
```

Then if the `BackendConfigFromParams` wanted to return a specific error code it would do:

```
  return nil, sqlproxyccl.NewErrorf(sqlproxyccl.CodeProxyRefusedConnection, ....)
```


<!-- Sent from Reviewable.io -->

spaskob reviewed (commented):
PTAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-:-MLsGA7a5FoUcVifaDiI:bb71pro)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @asubiotto, and @petermattis)

___
*[pkg/ccl/sqlproxyccl/metrics.go, line 76 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLrHfvbB24GMKfIK8VA:-MLsFJbj5r9J45riuN09:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5055720eab4f00ace41f8879997fea8c6142788/pkg/ccl/sqlproxyccl/metrics.go#L76)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/metaBackendDisconnectCount/metaRefusedConnCount/g`
</blockquote></details>

Done.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLDFJWH-8cFKYC8H56L:-MLsFbZl73Dw84qZJAaY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f373eb953bd93b22c0f0fdf8390d228e9b7f524b/pkg/ccl/sqlproxyccl/proxy.go#L59)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

Will do 
</blockquote></details>

Done.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 123 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNTuKx-RTJo4B97wTu:-MLsFgMoEYsgFDEBcZw4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/proxy.go#L123)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

Now that I understand better the
</blockquote></details>

Done.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 112 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-ML7Wk4mqJ6wzlfGHwA4-r5-112:-MLsFdH3Egx5UnpBHO9Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f079b2cb76155bcea7427fd5882afe6b5ba4d53/pkg/ccl/sqlproxyccl/proxy.go#L112)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Yes, that approach sounds good, though I have a nit about the naming. The "known tenants" cache is caching whether a tenant connection attempt was successful. Naming it `CacheConn` makes me think that it is caching the connection itself. I'd name this callback something like `ConnectionSuccess`.
</blockquote></details>

Done.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 28 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLrHE4U2RghXOr3aE_-:-MLsFKQ_4TN9hedZuRdg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5055720eab4f00ace41f8879997fea8c6142788/pkg/ccl/sqlproxyccl/proxy.go#L28)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Let's document these fields.
</blockquote></details>

Done.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 31 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLrJp0uDo28X4Xj_dbd:-MLsFLIb0s5Nm-MuZxP9:b-2ig2r8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5055720eab4f00ace41f8879997fea8c6142788/pkg/ccl/sqlproxyccl/proxy.go#L31)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Under what circumstances would this callback want to return an error?
</blockquote></details>

I was thinking on service timeout when this becomes a service one day but I guess we can hide all this complexity in the closure and only emit errors in the log as this is not really a serious failure that requires terminating
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 133 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLrI4WCCUM70DL6Vo6z:-MLsFeG67jUAVO_-6hIe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5055720eab4f00ace41f8879997fea8c6142788/pkg/ccl/sqlproxyccl/proxy.go#L133)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Rather than the special `backendConfig.RefuseConn` handling, perhaps we should export `newErrorf` and then allow inspecting it here. That is, we'd do something like:

```
if clientErr != nil {
  s.metrics.RoutingErrCount.Inc(1)

  var codeErr *codeError
  if !errors.As(clientErr, &codeErr) {
    codeErr = newErrorf(CodeParamsRoutingFailed, "rejected by BackendConfigFromParams: %v", clientErr)
  }
  sendErrToClient(conn, codeErr.code, clientErr.Error())
  return codeErr
}
```

Then if the `BackendConfigFromParams` wanted to return a specific error code it would do:

```
  return nil, sqlproxyccl.NewErrorf(sqlproxyccl.CodeProxyRefusedConnection, ....)
```
</blockquote></details>

Done.
___
*[pkg/ccl/sqlproxyccl/admitter/local.go, line 116 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLNKfkYALeEUcK-Bq6e:-MLsFGPW5t3vZMhrWgbM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/admitter/local.go#L116)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[pkg/ccl/sqlproxyccl/admitter/service.go, line 23 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-MLiEpFp9KFgawt6vDfg:-MLsFFAm53MxHCvpj6mS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df9b92f57c58894da60acf30a3369e8dbbb3b20e/pkg/ccl/sqlproxyccl/admitter/service.go#L23)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	// KnownClient checks if this client has connected successfully before.
> 	KnownClient(ipAddress string, tenID uint64) bool
> ```
Is it definitely better to have this as a separate method and force the caller to check this before calling `AllowRequest`? Should we just augment the `AllowRequest` interface to also take the `tenID` and short-circuit?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56192#-:-MLtKtu8AXWKLd1KWazz:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @asubiotto, and @petermattis)



<!-- Sent from Reviewable.io -->

spaskob commented:
TFTRs!

spaskob commented:
bors r+

spaskob commented:
bors r+

spaskob commented:
bors r+

spaskob commented:
bors r-

spaskob commented:
bors r+

spaskob commented:
bors r+
